tasks/ast_tools/src/generators/typescript.rs

```diff
@@ -0,0 +1,170 @@
+            Some(ty) => ty.clone(),
+            None => type_to_string(field.typ.name()),
+        };
+
+        if field.markers.derive_attributes.estree.flatten {
+            extends.push(ty);
+            continue;
+        }
+
+        let name = match &field.markers.derive_attributes.estree.rename {
+            Some(rename) => rename.to_string(),
+            None => field.name.clone().unwrap().to_case(Case::Camel),
+        };
+
+        fields.push_str(&format!("\n\t{name}: {ty};"));
+    }
+
+    let extends_union = extends.iter().any(|it| it.contains('|'));
+
+    if extends_union {
+        let extends =
+            if extends.is_empty() { String::new() } else { format!(" & {}", extends.join(" & ")) };
+        format!("export type {ident} = ({{{fields}\n}}){extends};")
+    } else {
+        let extends = if extends.is_empty() {
+            String::new()
+        } else {
+            format!(" extends {}", extends.join(", "))
+        };
+        format!("export interface {ident}{extends} {{{fields}\n}}")
+    }
+}
+
+fn type_to_string(ty: &TypeName) -> String {
+    match ty {
+        TypeName::Ident(ident) => match ident.as_str() {
+            "f64" | "f32" | "usize" | "u64" | "u32" | "u16" | "u8" | "i64" | "i32" | "i16"
+            | "i8" => "number",
+            "bool" => "boolean",
+            "str" | "String" | "Atom" | "CompactStr" => "string",
+            ty => ty,
+        }
+        .to_string(),
+        TypeName::Vec(type_name) => format!("Array<{}>", type_to_string(type_name)),
+        TypeName::Box(type_name) | TypeName::Ref(type_name) | TypeName::Complex(type_name) => {
+            type_to_string(type_name)
+        }
+        TypeName::Opt(type_name) => format!("({}) | null", type_to_string(type_name)),
+    }
+}
+
+fn format_typescript(source_text: &str) -> String {
+    let mut dprint = Command::new("dprint")
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .args(["fmt", "--stdin", "types.d.ts"])
+        .spawn()
+        .expect("Failed to run dprint (is it installed?)");
+
+    let stdin = dprint.stdin.as_mut().unwrap();
+    stdin.write_all(source_text.as_bytes()).unwrap();
+    stdin.flush().unwrap();
+
+    let output = dprint.wait_with_output().unwrap();
+    String::from_utf8(output.stdout).unwrap()
+}
+
+// Unusable until oxc_prettier supports comments
+// fn format_typescript(source_text: &str) -> String {
+//     let allocator = Allocator::default();
+//     let source_type = SourceType::ts();
+//     let ret = Parser::new(&allocator, source_text, source_type)
+//         .with_options(ParseOptions { preserve_parens: false, ..ParseOptions::default() })
+//         .parse();
+//     Prettier::new(
+//         &allocator,
+//         PrettierOptions {
+//             semi: true,
+//             trailing_comma: TrailingComma::All,
+//             single_quote: true,
+//             ..PrettierOptions::default()
+//         },
+//     )
+//     .build(&ret.program)
+// }
```

Comment:
Delete this dead code. Dprint is probably what we want to use anyway, since that's the tool we use to format all other JS/TS code in the repo.

Reply:
Only problem with dprint is it doesn't turn `(foo) | null` into `foo | null`.

Reply:
I was going to ask. Can we just output `foo | null` to start with?

I imagine in some cases `foo` *needs* wrapping in brackets, but I don't know TS well enough to know in what circumstances. Could we use a quick-and-dirty approach of wrapping only if it contains `{` or something?

Reply:
Yeah I went through and it doesn't seem to cause any issues here. If we were using prettier it would just be a no-brainer to include it but since dprint can't handle it it makes sense to remove the parentheses.

Reply:
I accidentally committed this to #6796, still getting used to graphite, sorry.

Reply:
No problem. It's just a nit anyway!